common/canonical_serialization/src/canonical_deserialize.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use failure::prelude::*;
+use std::collections::BTreeMap;
+
+pub trait CanonicalDeserialize {
+    fn deserialize(deserializer: &mut impl CanonicalDeserializer) -> Result<Self>
+    where
+        Self: Sized;
+}
+
+macro_rules! impl_canonical_deserializer {
+    ($(($function:ident, $type:ty),)+ $(tuple: ($tuple_function:ident, $($tuple_type:ident)+),)+) => (
+        pub trait CanonicalDeserializer {
+            $(fn $function(&mut self) -> Result<$type>;)+
+
+            $(
+                fn $tuple_function<$($tuple_type: CanonicalDeserialize,) +>(
+                    &mut self,
+                ) -> Result<($($tuple_type), +)>;
+            )+
+
+            // decode a byte array with the given length as input
```

Comment:
Am I missing something here?  It says decode byte array but the function is to decode a BTreeMap?